pkg/server/image_pull.go

```diff
@@ -420,3 +448,68 @@ func getLayers(ctx context.Context, key string, descs []imagespec.Descriptor, va
 	}
 	return
 }
+
+type inProgress struct {
+	entries sync.Map
+}
+
+func newInProgress() *inProgress {
+	return &inProgress{}
+}
+
+func (inp *inProgress) load(key string) (*imageBroadCaster, bool) {
+	val, ok := inp.entries.Load(key)
+	if ok {
+		return val.(*imageBroadCaster), true
+	}
+	return nil, false
+}
+
+func (inp *inProgress) store(key string) *imageBroadCaster {
+	bc := newImageBroadcaster()
+	inp.entries.Store(key, bc)
+	return bc
+}
+
+func (inp *inProgress) delete(key string) {
+	inp.entries.Delete(key)
+}
+
+type imageBroadCaster struct {
+	c     *sync.Cond
+	image containerd.Image
+	done  bool
+	err   error
+}
+
+func newImageBroadcaster() *imageBroadCaster {
+	return &imageBroadCaster{
+		c: sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (inp *inProgress) done(key string, err error, image containerd.Image) {
+	broadcaster, ok := inp.load(key)
+	if ok {
```

Comment:
is !ok ever a possibility? If not this condition should be surfaced somehow like an assert, panic or at least a log.

Reply:
Not unless something really went wrong, in which case I agree. I can add a log here.

Reply:
err, I don't know why I'm even loading here to begin with. We return the broadcaster in the store call anyways, so I should just pass it in this function and then we'd really never hit this lol.